Mathlib/Data/List/Shortlex.lean

```diff
@@ -0,0 +1,285 @@
+
+instance isAsymm [IsAsymm α r] : IsAsymm (List α) (Shortlex r) where
+  asymm := by
+    intro a b ab ba
+    rcases shortlex_def.mp ab with h1 | h2
+    · rcases shortlex_def.mp ba with h3 | h4
+      · omega
+      omega
+    rcases shortlex_def.mp ba with h3 | h4
+    · omega
+    exact List.Lex.isAsymm.aux r a b h2.2 h4.2
+
+theorem append_right {s₁ s₂ : List α} (t : List α) : Shortlex r s₁ s₂ →
+    Shortlex r s₁ (s₂ ++ t) := by
+  intro h
+  rcases shortlex_def.mp h with h1 | h2
+  · apply of_length_lt
+    rw [List.length_append]
+    omega
+  cases t with
+  | nil =>
+    rw [List.append_nil]
+    exact h
+  | cons head tail =>
+    apply of_length_lt
+    rw [List.length_append, List.length_cons]
+    omega
+
+theorem append_left {t₁ t₂ : List α} (h : Shortlex r t₁ t₂) (s : List α) :
+    Shortlex r (s ++ t₁) (s ++ t₂) := by
+  rcases shortlex_def.mp h with h1 | h2
+  · apply of_length_lt
+    rw [List.length_append, List.length_append]
+    omega
+  cases s with
+  | nil =>
+    rw [List.nil_append, List.nil_append]
+    exact h
+  | cons head tail =>
+    apply of_lex
+    · simp only [List.cons_append, List.length_cons, List.length_append, Nat.succ_eq_add_one,
+      add_left_inj, add_right_inj]
+      exact h2.1
+    exact List.Lex.append_left r h2.2 (head :: tail)
+
+section WellFounded
+
+variable {h : WellFounded r}
+
+theorem shortlexAccessible {a : α} (n : ℕ) (aca : Acc r a)
+    (acb : (b : List α) → b.length < n → Acc (Shortlex r) b) (b : List α) (hb : b.length < n)
+    (ih : ∀ s : List α, s.length < (a::b).length → Acc (Shortlex r) s) :
+    Acc (Shortlex r) ([a] ++ b) := by
+  induction aca generalizing b with
+  | intro xa _ iha =>
+    induction (acb b hb) with
+    | intro xb _ ihb =>
+      apply Acc.intro ([xa] ++ xb)
+      intro p lt
+      rcases shortlex_def.mp lt with h1 | h2
+      · exact ih _ h1
+      · cases p with
+        | nil => simp only [List.length_nil, List.singleton_append, List.length_cons,
+          Nat.succ_eq_add_one, self_eq_add_left, add_eq_zero, List.length_eq_zero, one_ne_zero,
+          and_false, false_and] at h2
+        | cons headp tailp =>
+          cases h2.2 with
+          | cons h =>
+            rw [List.append_eq, List.nil_append] at h
+            simp only [List.length_cons, Nat.succ_eq_add_one, List.singleton_append,
+              add_left_inj] at h2
+            rw [← h2.1] at hb
+            apply ihb _ (of_lex (h2.1) h) hb
+            intro l hl
+            apply ih
+            rw [List.length_cons, ← h2.1]
+            exact hl
+          | rel h =>
+            simp only [List.length_cons, Nat.succ_eq_add_one, List.singleton_append,
+              add_left_inj] at h2
+            rw [← h2.1] at hb
+            apply iha headp h _ hb
+            intro l hl
+            apply ih
+            rw [List.length_cons, ← h2.1]
+            exact hl
+
+theorem wf {α : Type*} (r : α → α → Prop) {h : WellFounded r} : WellFounded (Shortlex r) := by
```

Comment:
```suggestion
theorem wf (h : WellFounded r) : WellFounded (Shortlex r) := by
```
since there is no way to infer `h` in your version, and the other two variables are already declared above.